### ISSUE TITLE
Overriding `__new__` with `Any` return annotation on a class's metaclass results in class instance resolving as `Any`

### ISSUE THREAD

**Issue by Mardoxx** [microsoft/pyright#5223]
**Describe the bug**
Overriding `__new__` with a return annotation of `Any` in a class' metaclass results in the type of the class instance resolving as `Any`.

**To Reproduce**
See code example

**Expected behavior**
`a` should be `MyClass` instead of `Any`

**Screenshots or Code**
```python
class MyMeta(type):
    def __call__(cls, *args: Any, **kwargs: Any) -> Any:
        # Do other stuff...
        return super().__call__(args, kwargs)
    
class MyClass(metaclass=MyMeta):
    ...

a = MyClass() # A is now "Any" instead of "MyClass"
```

**VS Code extension or command-line**
Pylance v2023.5.50

**Additional context**
Workaround is to add the following after the metaclass definition

```python
if typing.TYPE_CHECKING:
    class MyMeta(type): ...
```

You would need a ` # pyright: ignore[reportGeneralTypeIssues]` or equivalent on the class itself

**Comment by erictraut** [comment]
Yes, this is by design. If you explicitly specify that the metaclass `__call__` method returns `Any`, then pyright will honor this. If your intent is to return an instance of the class passed to the `__call__` method, you should use a TypeVar.

```python
T = TypeVar("T")

class MyMeta(type):
    def __call__(cls: type[T], *args: Any, **kwargs: Any) -> T: ...
```

**Comment by erictraut** [comment]
I've decided to revisit the design based on input from multiple pyright users. I've adjusted the heuristics so a metaclass `__call__` override that returns an `Any` will prefer the return type of the class `__new__` method rather than `Any`. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.320, which I just published. It will also be included in a future version of pylance.
